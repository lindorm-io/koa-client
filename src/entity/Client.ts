import { ClientEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  extra: TObject<any>;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  extra?: TObject<any>;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _extra: TObject<any>;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._extra = options?.extra || {};
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get extra(): TObject<any> {
    return this._extra;
  }
  public set extra(object: TObject<any>) {
    this._extra = object;
    this.addEvent(ClientEvent.EXTRA_CHANGED, { extra: this._extra });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new EntityCreationError("Client");
    }
    this.addEvent(ClientEvent.CREATED, {
      approved: this._approved,
      description: this._description,
      extra: this._extra,
      name: this._name,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
