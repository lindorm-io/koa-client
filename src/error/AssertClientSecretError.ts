import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertClientSecretError extends APIError {
  constructor(clientId: string, error?: Error) {
    super("Invalid Client Secret", {
      debug: { id: clientId, error },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
