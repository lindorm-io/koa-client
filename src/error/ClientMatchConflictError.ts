import { APIError, HttpStatus } from "@lindorm-io/core";

export class ClientMatchConflictError extends APIError {
  constructor(expect: string, actual: string) {
    super("Client Match Conflict", {
      debug: { expect, actual },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
