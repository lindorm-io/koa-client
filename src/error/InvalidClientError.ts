import { APIError, HttpStatus } from "@lindorm-io/core";

export class InvalidClientError extends APIError {
  constructor(clientId: string, error?: Error) {
    super("Invalid Client ID", {
      debug: { id: clientId, error },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
