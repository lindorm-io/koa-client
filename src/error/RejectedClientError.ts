import { APIError, HttpStatus } from "@lindorm-io/core";

export class RejectedClientError extends APIError {
  constructor(clientId: string, error?: Error) {
    super("Client is not approved", {
      debug: { id: clientId, error },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
