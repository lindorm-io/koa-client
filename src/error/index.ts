export * from "./AssertClientSecretError";
export * from "./ClientMatchConflictError";
export * from "./InvalidClientError";
export * from "./RejectedClientError";
