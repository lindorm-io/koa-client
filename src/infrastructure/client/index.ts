export * from "./ClientCache";
export * from "./ClientRepository";
