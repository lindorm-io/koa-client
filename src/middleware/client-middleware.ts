import { IKoaClientContext } from "../types";
import { InvalidClientError, RejectedClientError } from "../error";
import { TPromise } from "@lindorm-io/core";
import Joi from "@hapi/joi";

const schema = Joi.object({
  clientId: Joi.string().guid().required(),
});

export const clientMiddleware = async (ctx: IKoaClientContext, next: TPromise<void>): Promise<void> => {
  const start = Date.now();

  const { cache, logger, metadata } = ctx;
  const { clientId } = metadata;

  await schema.validateAsync({ clientId });

  try {
    ctx.client = await cache.client.find(clientId);
  } catch (err) {
    throw new InvalidClientError(clientId, err);
  }

  if (!ctx.client.approved) {
    throw new RejectedClientError(clientId);
  }

  logger.debug("valid client found", { clientId });

  ctx.metrics = {
    ...(ctx.metrics || {}),
    client: Date.now() - start,
  };

  await next();
};
