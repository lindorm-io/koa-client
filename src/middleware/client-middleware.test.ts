import MockDate from "mockdate";
import { Client } from "../entity";
import { ClientCache } from "../infrastructure/client";
import { IRedisConnectionOptions, RedisConnection, RedisConnectionType } from "@lindorm-io/redis";
import { InvalidClientError, RejectedClientError } from "../error";
import { Logger, LogLevel } from "@lindorm-io/winston";
import { TPromise } from "@lindorm-io/core";
import { clientMiddleware } from "./client-middleware";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

const redisOptions: IRedisConnectionOptions = {
  type: RedisConnectionType.MEMORY,
  port: 12345,
};

const logger = new Logger({ packageName: "n", packageVersion: "v" });
logger.addConsole(LogLevel.ERROR);

describe("clientMiddleware", () => {
  let ctx: any;
  let next: TPromise<void>;
  let client: Client;

  beforeEach(async () => {
    const redis = new RedisConnection(redisOptions);
    await redis.connect();

    client = new Client({ approved: true });
    ctx = {
      cache: {
        client: new ClientCache({
          client: redis.getClient(),
          logger,
        }),
      },
      logger,
      metadata: {
        clientId: client.id,
      },
    };
    next = () => Promise.resolve();

    await ctx.cache.client.create(client);
  });

  test("should set client on context", async () => {
    await expect(clientMiddleware(ctx, next)).resolves.toBe(undefined);

    expect(ctx.client).toMatchSnapshot();
    expect(ctx.metrics.client).toStrictEqual(expect.any(Number));
  });

  test("should reject client when id is mismatched", async () => {
    ctx.metadata.clientId = "e181c870-dd78-4d49-b9f5-9202f63f6e00";

    await expect(clientMiddleware(ctx, next)).rejects.toThrow(expect.any(InvalidClientError));
  });

  test("should reject client when it is not approved", async () => {
    client.approved = false;
    await ctx.cache.client.update(client);

    await expect(clientMiddleware(ctx, next)).rejects.toThrow(expect.any(RejectedClientError));
  });
});
