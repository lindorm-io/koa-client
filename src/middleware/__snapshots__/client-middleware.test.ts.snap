// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clientMiddleware should set client on context 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": null,
  "_events": Array [],
  "_extra": Object {},
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": null,
  "_secret": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
